include/irr/asset/filters/CSwizzleableAndDitherableFilterBase.h

```diff
@@ -17,20 +17,124 @@ namespace irr
 	{
 		namespace impl
 		{
+			namespace detail
+			{
+				union NormalizeValues
+				{
+					NormalizeValues() {}
+					~NormalizeValues() {}
+
+					NormalizeValues(NormalizeValues& copy)
+					{
+						std::memmove(this, &copy, sizeof(NormalizeValues));
+					}
+
+					NormalizeValues(const NormalizeValues& copy)
+					{
+						std::memmove(this, &copy, sizeof(NormalizeValues));
+					}
+
+					NormalizeValues& operator=(NormalizeValues& copy)
+					{
+						std::memmove(this, &copy, sizeof(NormalizeValues));
+						return *this;
+					}
+
+					NormalizeValues& operator=(const NormalizeValues& copy)
+					{
+						std::memmove(this, &copy, sizeof(NormalizeValues));
+						return *this;
+					}
+
+					core::vectorSIMDu32 f;
+					core::vectorSIMDu32 u;
+					core::vectorSIMDi32 i;
+				};
+
+				//! Normalizing state
+				/*
+					The class provides mininum and maximum values per texel
+					that may be used for normalizing texels.
+				*/
+
+				template<bool Normalize>
+				class CNormalizeState
+				{
+					public:
+						CNormalizeState() {}
+						virtual ~CNormalizeState() {}
+
+						NormalizeValues oldMaxValue, oldMinValue;
+
+					protected:
+
+						/*
+							The function normalizes curret texel passed to the function
+							using oldMaxValue and oldMinValue provided by the state.
+
+							It's a user or a programmer responsibility to update those
+							variables to contain proper values for normalizing the texel.
+
+							The function supports compile-time normalize.
+						*/
+
+						template<E_FORMAT format, typename Tenc>
+						void normalize(Tenc* encodeBuffer, const uint8_t& channels)
+						{
+							if constexpr (isSignedFormat<format>())
+								for (uint8_t channel = 0; channel < channels; ++channel)
+									encodeBuffer[channel] = (2.0 * encodeBuffer[channel] - oldMaxValue.f[channel] - oldMinValue.f[channel]) / (oldMaxValue.f[channel] - oldMinValue.f[channel]);
+							else
+								for (uint8_t channel = 0; channel < channels; ++channel)
+									encodeBuffer[channel] = (encodeBuffer[channel] - oldMinValue.f[channel]) / (oldMaxValue.f[channel] - oldMinValue.f[channel]);
+						}
+
+						/*
+							Runtime version of normalize
+
+							@see normalize
+						*/
+
+						template<typename Tenc>
+						void normalize(const E_FORMAT& format, Tenc* encodeBuffer, const uint8_t& channels)
+						{
```

Comment:
`static_assert` and `assert` that the format is not a pure-integer format

Reply:
corrected in following commit